### PR TITLE
[CUDA][SDPA] Don't dispatch to mem eff attn for batch_size >= 65536

### PR DESCRIPTION
**Pull request by eqy**
#146704

cc @ptrblck @msaroufim

### CODE REVIEW

**Comment by nikitaved** [comment]
The operation, unless the kernel is fixed in terms of the grid handling, is separable in the batch dim. So why not just cover the batch dims with several kernel launches?

File: test/test_transformers.py [hunk 1]
```diff
@@ -1899,6 +1899,16 @@ def test_flash_attention_fail_with_non_square_causal_attention(self, device):
                 self.assertRaises(RuntimeError, lambda: torch.nn.functional.scaled_dot_product_attention(
                     q, k, v, None, 0.0, is_causal=True))
 
+    @onlyCUDA
+    def test_mem_eff_attention_fail_with_batch_size_geq_65536(self, device):
+          query = torch.rand([2**16, 8, 4, 64], device='cuda', dtype=torch.float16)
```
Comment (drisspg) [inline test/test_transformers.py:1904]: I wonder if this will OOm on runners?

**Review by drisspg — APPROVED** [review]

File: test/test_transformers.py (hunk 1 above)
Comment (nikitaved) [reply test/test_transformers.py:1904]: The spatial dims could be dialed down. Is this a temporary fix, or the user would need to split tensors?

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 2]
```diff
@@ -289,6 +289,18 @@ bool check_mem_efficient_hardware_support(sdp_params const& params, bool debug)
   return true;
 }
 
+bool check_mem_efficient_batch_size(sdp_params const& params, bool debug) {
```
Comment (Skylion007) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:292]: Doesn't mem efficient attention live in the PyTorch cudabase + xformers? Should we just fix this properly by changing a dtype / type somewhere?

Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:292]: mem-efficient lives entirely within PyTorch code base. I think the problem isn't dtype related rather that cuda imposes limits on a singular grid dim size. 

The proper fix would likely be to split this large grid dim among another

Comment (Skylion007) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:292]: Or just launching two kernels, right? It's separable along batch.

Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:292]: yeah or this
